src/privatesend-client.cpp

```diff
@@ -963,14 +973,37 @@ bool CPrivateSendClient::StartNewQueue(CAmount nValueMin, CAmount nBalanceNeedsA
             nSessionDenom = CPrivateSend::GetDenominationsByAmounts(vecAmounts);
         }
 
+        // Count available denominations.
+        // Should never really fail after this point, since we just selected compatible inputs ourselves.
+        std::vector<int> vecBits;
+        if (!CPrivateSend::GetDenominationsBits(nSessionDenom, vecBits)) {
+            return false;
+        }
+
+        CAmount nValueInTmp = 0;
+        std::vector<CTxDSIn> vecTxDSInTmp;
+        std::vector<COutput> vCoinsTmp;
+        std::vector<CAmount> vecStandardDenoms = CPrivateSend::GetStandardDenominations();
+
+        bool fSelected = pwalletMain->SelectCoinsByDenominations(nSessionDenom, vecStandardDenoms[vecBits.front()], vecStandardDenoms[vecBits.front()] * PRIVATESEND_ENTRY_MAX_SIZE, vecTxDSInTmp, vCoinsTmp, nValueInTmp, 0, nPrivateSendRounds);
```

Comment:
Here and in `JoinExistingQueue` you do selection for whole range of PS rounds. And you assume this amount of inputs to set `nInputCount` (or to check in `JoinExistingQueue`)

But the more value of `nInputCount`, the more likely failure of [this code](https://github.com/dashpay/dash/blob/84f61ca0db2ef60527bf34448537957c0dc41a98/src/privatesend-client.cpp#L1043-L1062), since in `PrepareDenominate` selection is limited to only one PS round.

Reply:
Yes, that's a good point but the code that follows the one you pointed to covers such cases, so shouldn't be an issue imo.

Reply:
I know, but now mixing will be complete mess:
- in some cases will be selected inputs with highest number of rounds
- in some cases will be selected inputs with different(!) number of rounds

Reply:
Ah, I see what you mean. But the former one is just the optimal strategy to get some PS balance quicker, while the later one is still a good one because it either advances low number inputs (which will help to get some more PS balance later) or mixes inputs which already have enough rounds and this shouldn't hurt either.

Reply:
Or am I missing some another/new vulnerability here?

Reply:
I don't think it's vulnerability. However I see possible scenario which reduce privacy.

Preconditions: every mixing transaction have equal number of inputs/outputs from each participant and inputs for one mixing transaction can be from different rounds.

[See this diagram](https://i.imgur.com/uRvegkN.png) (your inputs/outputs are blue) 

You've done mixing in tx1. Then you joined mixing tx2 and you used only 2 outputs from previous round (tx1). Then you joined mixing tx3, but now you have not enough outputs from previous round (tx2) and you using outputs from different rounds (from tx1 and tx2). 

Since I know that every participant in mixing transaction tx3 have 3 inputs and I see that other 2 participants each have 3 inputs from other mixing transactions (txY and txZ) I can link your inputs and drop your one mixing round (tx2) from analysis.

Reply:
For the best possible privacy, the inputs likely should be selected randomly. Any other setup has the potential for data leakage and trace-ability, maybe there is a nice way to make it look random while also maintaining speed of getting a balance. However, the random setup would also have a problem for single input private-sending, unless there was very high liquidity of single input only mixing. Maybe we need to say you can't only mix one input?

However, having it be that "in some cases inputs will be selected with highest number of rounds" is likely a very bad idea as there will likely be grouping (when you look at Redeemed in and previous output) happening allowing for the breaking of a mix.

Reply:
I agree with PaulieD. If the inputs were selected randomly then I think you couldn't really deduce this:
> other 2 participants each have 3 inputs from other mixing transactions (txY and txZ)
. But if the most recent inputs are always used then I think that InhumanPerfection is correct. 

I don't really understand the original issue. If someone could come up with an example scenario where the privacy is reduced by participants using a different number of inputs, then I might be able to understand it. The way I see it currently is that since a blockchain observer shouldn't be able to know which inputs belong to which mixing participant, then the observer also shouldn't be able to tell how many inputs belong to each participant.

Reply:
@PaulieD Do you think prioritizing inputs with highest number of rounds might have also caused #2092 or made it significantly more common?

Reply:
I can confirm that there is a definite problem that this pr should fix. @UdjinM6 might be able to provide more details.

Reply:
@Antti-Kaikkonen yeah, if the selections were random you certainly wouldn't see that